wazo_call_logd/tests/test_generator.py

```diff
@@ -151,3 +171,240 @@ def _generate_cel_for_call(self, linked_id, cel_count=3):
+        self._users = users or []
+        self._extensions = extensions or []
+        self._context = context
+        self._tenant_uuid = tenant_uuid
+
+    def id_(self):
+        return self._id
+
+    def users(self):
+        return self._users
+
+    def to_dict(self):
+        return {
+            'id': self._id,
+            'name': self._name,
+            'protocol': self._protocol,
+            'context': self._context,
+            'tenant_uuid': self._tenant_uuid,
+            'users': [
+                u.to_dict()
+                for u in self._users
+            ],
+            'extensions': self._extensions,
+        }
+
+
+class MockSwitchboard:
+    def __init__(self, uuid, name=None):
+        self._uuid = str(uuid)
+        self._name = name
+
+    def uuid(self):
+        return self._uuid
+
+    def to_dict(self):
+        return {'uuid': self._uuid, 'name': self._name}
+
+
+class MockContext:
+    def __init__(self, id, name, tenant_uuid):
+        self._id = id
+        self._name = name
+        self._tenant_uuid = str(tenant_uuid)
+
+    def id_(self):
+        return self._id
+
+    def to_dict(self):
+        return {'id': self._id, 'name': self._name, 'tenant_uuid': self._tenant_uuid}
+
+
+def parse_cel_table(cel_table: str):
+    cels = []
+    lines = cel_table.strip().split('\n')
+    columns = [field.strip() for field in lines[0].split('|')]
+    for line in lines[2:]:
+        cel_fields = [field.strip() for field in line.split('|')]
+        cel = CEL(**dict(zip(columns, cel_fields)))
+        cels.append(cel)
+    return cels
+
+
+def setup_confd_client(confd_client, users=None, lines=None, contexts=None):
+    users_map = {
+        user.uuid: user.to_dict()
+        for user in users
+    } if users else {}
+    print("users_map: ", users_map)
+    confd_client.users.get.side_effect = (lambda uuid: users_map[uuid])
+    confd_client.lines.list.side_effect = lines and (lambda **kwargs: dict(
+        items=[
+            line.to_dict()
+            for line in lines
+        ]
+    ))
+    confd_client.contexts.list.side_effect = contexts and (lambda **kwargs: dict(
+        items=[
+            context.to_dict()
+            for context in contexts
+        ]
+    ))
+
+
+class TestCallLogsGeneration(TestCase):
+    def setUp(self):
+        self.confd_client = create_autospec(
+            ConfdClient, instance=True, 
+            users=create_autospec(users.UsersCommand, instance=True),
+            contexts=create_autospec(contexts.ContextsCommand, instance=True),
+            lines=create_autospec(lines.LinesCommand, instance=True)
+        )
+        self.generator = CallLogsGenerator(
+            self.confd_client,
+            [
+                LocalOriginateCELInterpretor(),
+                DispatchCELInterpretor(
+                    CallerCELInterpretor(),
+                    CalleeCELInterpretor(),
+                ),
+            ],
+        )
+
+    def test_unreachable_user(self):
```

Comment:
Having experienced frustration with debugging through integration tests, this is a proof of concept for having functional testing as unit tests to enable a tighter, better test loop and developer experience.
The higher level of complexity and indirection involved in integration tests makes it hard to debug simple errors in the core logic with them.

This example may be more verbose than ideal(though not much more than integration tests), some improvement could be made to that.